### PR TITLE
feat: remove ipld legacy and format globals in ipld/merkledag

### PR DESCRIPTION
**Pull request by aschmahmann**
Related to https://github.com/ipfs/boxo/issues/291

### CODE REVIEW

File: ipld/merkledag/merkledag.go
Comment (aschmahmann) [inline ipld/merkledag/merkledag.go:25]: @willscott WDYT about explicitly passing the go-ipld-prime linksystem here instead of requiring the default? This might make it easier for people to get out of the "global defaults" game.

File: ipld/merkledag/merkledag.go
Comment (aschmahmann) [inline ipld/merkledag/merkledag.go:28]: @willscott thanks for doing https://github.com/ipfs/go-merkledag/pull/104. I mostly copied it. Although I left the globals up in the global space as a reminder for us to try and exit globals land when it becomes possible 😅

File: ipld/merkledag/merkledag.go
Comment (willscott) [reply ipld/merkledag/merkledag.go:25]: is there a reason to have this as a global versus the approach i took in https://github.com/ipfs/go-merkledag/pull/104 and making the decoder in the `NewDAGService` below?

**Comment by Jorropo** [comment]
https://github.com/ipfs/kubo/pull/9787#issuecomment-1532313618 :rofl: #309 has some merits I see

File: ipld/merkledag/merkledag.go
Comment (willscott) [reply ipld/merkledag/merkledag.go:28]: > as a reminder for us to try and exit globals land when it becomes possible 

I'm confused by this - in 104 i did exit globals land. - is there a reason not to stay consistent with that in the copy here, since it's the direction we want to go anyway?

Comment (aschmahmann) [reply ipld/merkledag/merkledag.go:28]: > I'm confused by this - in 104 i did exit globals land

We're still stuck with globals:
1. The merkledag libraries still have globals. IMO this is bad for the same reasons the other ones were bad, it just turns out the other ones are basically unused independently so for now we can at fairly low cost coalesce here (and solve a problem we currently have)
2. There are still go-ipld-prime globals used implicitly in go-ipld-legacy unless we fix that https://github.com/ipfs/go-ipld-legacy/pull/12/files#r1211104340
    - Ran out of time to do the kubo PR exercising this (https://github.com/ipfs/kubo/pull/9907). Maybe it'll be fine to leave this merkledag PR mostly alone (aside from passing in something like a linksystem to go-ipld-legacy). However, to start wrestling control away from the globals we might also want to make the legacy decoder used by merkledag publicly settable (albeit with the various warnings about being careful about using it, and trying to only use it in top-level binaries).
    
>  is there a reason not to stay consistent with that in the copy here

If we go with making the go-ipld-legacy decoder publicly settable (and therefore not tied to go-ipld-prime's global registry) then this approach looks much better rather than just being a stylistic difference. If it's just a stylistic difference then 🤷.

File: ipld/merkledag/merkledag.go
Comment (aschmahmann) [reply ipld/merkledag/merkledag.go:25]: Initially it was just a stylistic preference that I had since a) it's still a global in 104, it's just a little more subtle b) it's a bit of a DRY thing where I didn't want to accidentally change one instead of the other.

However, I'm now also thinking that given we still have this global we might want to make it settable until we can make relying on a merkledag global registry no longer a thing (especially since it also implicitly hides the go-ipld-prime global registry)

**Review by willscott — APPROVED** [review]

**Comment by aschmahmann** [comment]
merging without approval from the other kubo maintainers since I feel like @willscott's review is enough 😄. If someone objects we can fix it up before the boxo release tag.

**Pull request merged**
